engine/source_sym_engine.ml

```diff
@@ -0,0 +1,184 @@
+open Ast
+
+type constraint_tree =
+  | Leaf of pi list * source_expr
+  | Node of pi * constraint_tree list
```

Comment:
I'm not sure why we store `pi` constraints both at the leaves and in the nodes. I would assume that either we store the constraints along each path in the corresponding leaf (so no need for pi on the nodes, but also no sharing of constraint information along several paths), or we store the constraints in the tree nodes (would have my preference), and then we probably don't need additional constraints on the leaves?